### PR TITLE
Implement Flyouts

### PR DESCRIPTION
**Pull request by amwx**
## What does the pull request do?

Add Flyout implementation as in UWP/WinUI (#5172)

Questions from #5172
- Should ContextMenu be deprecated in favor of Flyouts (i.e., ContextFlyout property), as in UWP/WinUI. This would be a breaking change. (personally I'd be in favor of this)
- Integration with Native Menus (I'm not including this here as I'm not too familiar with native menus and don't have a way to test them anyway, since they don't work on Windows iirc)
- Open question 2: should some (or all) of the popup customization options (like offsets, etc) on Popup be visible through flyout, which differs from WinUI API. I've gotten most of the FlyoutPlacementMode options working, which WinUI has many, so that may make this unnecessary.

Since FlyoutBase is an `AvaloniaObject`, it itself is really just a more managed Popup. One difference between UWP/WinUI and Avalonia, is WinUI allows you to still name Flyouts (x:Name="Flyout") and target them through their styling system. Since Flyout is not a `StyledElement` this isn't possible in Avalonia, thus you can't, for example, change the Placement mode in a separate `Style` entry if using it in a Template.

To Implement/Checklist:
- [x] FlyoutBase
- [x] Flyout
- [x] FlyoutPresenter
- [x] Default FlyoutPresenter Style for Fluent Theme
- [x] Default FlyoutPresenter Style for Default Theme
- [x] Flyout property for Button
- [x] MenuFlyout
- [x] Tests
- [x] Sample page

*API differences from WinUI*
- `MenuFlyout` uses existing `MenuBase` logic, and will essentially work like the old content menu, and should ease the transition. `MenuFlyoutItem`, `MenuFlyoutSubitem`, `MenuFlyoutSeparator` were not ported, continue to use `MenuItem` as usual
- `FlyoutShowOptions` and `FlyoutBase.ShowAt(Control, FlyoutShowOptions)` was not ported

**Breaking Change**
ContextMenu property has been marked obsolete in favor of ContextFlyout

### CODE REVIEW

**Comment by robloo** [comment]
Excited to see this come together. When it's ready I'm going to use it to implement #4960 and #4959. I don't think there will be any significant roadblocks after this.

**Comment by kekekeks** [comment]
> Since Flyout is not a StyledElement this isn't possible in Avalonia

IIRC styling won't be tied to visuals soon-ish. That's required for the flow document port.

File: src/Avalonia.Controls/Button.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Button.cs:85]: ContextMenu and ContextFlyout are StyledProperty, so it's possible to set them from styles and have multiple prioritized values.
Why Button.Flyout is simple DirectProperty?

Comment (maxkatz6) [reply src/Avalonia.Controls/Button.cs:85]: What if flyout was changed while it was opened? Not only on button, but also ContextFlyout one too.
I don't know actually, how it handled in UWP now.

File: src/Avalonia.Controls/Button.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Button.cs:313]: What if flyout was changed while it was opened? Not only on button, but also ContextFlyout one too.
I don't know actually, how it handled in UWP now.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/FlyoutBase.cs:137]: Should "canCancel" be part of the public API?
See: [https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.primitives.flyoutbase.hide?view=winrt-19041#Windows_UI_Xaml_Controls_Primitives_FlyoutBase_Hide](docs)

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/FlyoutBase.cs:421]: It's better to not include not supported enum values in the API.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/FlyoutBase.cs:440]: Keyboard "Menu" key should be also supported.

File: src/Avalonia.Controls/Flyouts/MenuFlyoutPresenter.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/MenuFlyoutPresenter.cs:27]: Probably you can save MenuFlyout instance inside of MenuFlyoutPresenter, when it's created.
And show this flyout from parent logical element, if it exists.
But I am not sure in what cases this method can be called @grokys .

ContextMenu tries to implement this method using saved list of attached controls https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/ContextMenu.cs#L255

File: src/Avalonia.Themes.Default/FlyoutPresenter.xaml
Comment (maxkatz6) [inline src/Avalonia.Themes.Default/FlyoutPresenter.xaml:2]: Default Flyout style should have some padding too, otherwise it doesn't look good out of box.
Maybe 4px?

File: src/Avalonia.Themes.Fluent/Controls/FlyoutPresenter.xaml
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Controls/FlyoutPresenter.xaml:42]: Maybe we should have CornerRadius property on FlyoutPresenter from the beginning and use it the templates with TemplateBinding?

File: src/Avalonia.Themes.Fluent/Controls/TextBox.xaml
Comment (maxkatz6) [inline src/Avalonia.Themes.Fluent/Controls/TextBox.xaml:29]:
```suggestion
    </ContextMenu>
```

File: src/Avalonia.Controls/Flyouts/MenuFlyoutPresenter.cs
Comment (maxkatz6) [reply src/Avalonia.Controls/Flyouts/MenuFlyoutPresenter.cs:27]: Or you can use FindLogicalAncestorOfType<Popup> in same way, as it was used here in Close method

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/FlyoutBase.cs:458]: Should we check if ContextMenu was set on the control, and ignore ContextFlyout if it was? To avoid breaking changes (let's say, user has set custom ContextFlyout on the text box, and with new update app will show two popups or only new one).
Maybe some warning in the user logs will be useful here as well.

File: src/Avalonia.Controls/Flyouts/MenuFlyout.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/MenuFlyout.cs:15]: Is MenuFlyout placement the same as in FlyoutBase be default? I thought it is "Bottom".

**Comment by maxkatz6** [comment]
> Should ContextMenu be deprecated in favor of Flyouts (i.e., ContextFlyout property), as in UWP/WinUI. This would be a breaking change. (personally I'd be in favor of this)

Yes. And it most likely not possible to reuse Flyout logic or MenuFlyoutPresenter styles inside of ContextMenu without another breaking changes.

> Integration with Native Menus (I'm not including this here as I'm not too familiar with native menus and don't have a way to test them anyway, since they don't work on Windows iirc)

Yes, it's not required to be implemented now.

> Open question 2: should some (or all) of the popup customization options (like offsets, etc) on Popup be visible through flyout, which differs from WinUI API. I've gotten most of the FlyoutPlacementMode options working, which WinUI has many, so that may make this unnecessary.

It can be discussed and implemented later.
Probably it can be done with attached properties to keep API simpler and avoid conflicts with native menus in the future, but there wasn't any agreement AFAIK.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/FlyoutBase.cs:231]: This method implementation is hard to understand without comments.
Anyway, it seems to work fine with mouse, but keyboard is not handled in any way (does UWP handle it?).

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (robloo) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:421]: As this enum is carry-over from UWP should an exception be made in this case? Keeping the enum value in now ensures binary compatiblity at least so it's a form of future-proofing the design. It's entirely possible 'Full' is needed in the future (actually it's really useful for dialogs).

File: src/Avalonia.Themes.Fluent/Controls/FlyoutPresenter.xaml
Comment (robloo) [reply src/Avalonia.Themes.Fluent/Controls/FlyoutPresenter.xaml:42]: Definitely would like to see CornerRadius make it's way into controls.

**Comment by robloo** [comment]
> MenuFlyout uses existing MenuBase logic, and will essentially work like the old content menu, and should ease the transition. MenuFlyoutItem, MenuFlyoutSubitem, MenuFlyoutSeparator were not ported, continue to use MenuItem as usual

Is there any plan to implement MenuFlyoutItem, MenuFlyoutSubitem, MenuFlyoutSeparator in Avalalonia in the future? 

MenuItem from WPF seems like a better design to me (don't know why Subitem exists in UWP). It seems like this flyout implementation actually hits the sweet spot using good ideas from both WPF and UWP. The only downside it is now is not API compatible with either.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (kekekeks) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:421]: Binary compatibility with what? UWPs enum? If we really wish to have that, we could leave a hole in the enumeration.

**Comment by maxkatz6** [comment]
> Is there any plan to implement MenuFlyoutItem, MenuFlyoutSubitem, MenuFlyoutSeparator in Avalalonia in the future?

There was issue somewhere with MenuItem improvements. We need at least ToggleMenuItem (it's possible to use toggling menuitem now, but not in convenient way) and RadioMenuitem. And probably these types should not have subitems list property.

**Comment by robloo** [comment]
> There was issue somewhere with MenuItem improvements. We need at least ToggleMenuItem (it's possible to use toggling menuitem now, but not in convenient way) and RadioMenuitem. And probably these types should not have subitems list property.

Oh yes, thanks for reminding me why the 'MenuFlyoutSubitem' exists (So child items are not included with the other menu item types like toggle/radio).

With that in mind it might be better to go the UWP API route (but definitely would prefer 'Flyout' wasn't included in the MenuItem names). That said, it's probably not work a breaking change in names so full UWP design might be best after all.

This might be the other issue: https://github.com/AvaloniaUI/Avalonia/issues/224

**Comment by maxkatz6** [comment]
@robloo after my comment I was thinking, that someone might actually want to have toggleable menuitem with sub items. In case of toggle buttons it can be like:
root:
> feature1 (enable/disable)
> feature2 (enable/disable)
- > feature 2 option (enable/disable)
- > feature 2 option (enable/disable)

But it's not directly related to Flyout feature, so I think it can be discussed in https://github.com/AvaloniaUI/Avalonia/issues/224 or new discussion later.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (robloo) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:421]: Yes, with the UWP enum and any apps that might be using the int value itself for whatever reason. Would that affect many real-world apps? Probably not. I've just been conditioned to avoid making changes to enums especially in the middle (Full = 4). If this was at the end of the enum it wouldn't matter or if there was never any chance that Full would be supported in Avalonia. It just seems like a necessary future-proofing change to me.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:137]: Probably not. Will fix

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:421]: I would've like to implement `Full` but the positioning API for `Popup` is limited in that sense. It's the same reason I didn't include the `FlyoutShowOptions` which has an ExclusionRect (a Rect the popup tries to avoid displaying in), and a Point property (to show the Flyout at a specific point). This would all probably be doable with Overlay popups, but when their windowed it would require getting creative. 
For now, I'll just leave a hole in the enum where Full is.

File: src/Avalonia.Controls/Flyouts/MenuFlyoutPresenter.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/MenuFlyoutPresenter.cs:27]: `Open` is abstract from `MenuBase` so it had to be included, but from what I can tell it was never called from `DefaultMenuInteractionHandler`, so I think the only way would be if you called it directly. IMO, ignoring should be ok as one normally shouldn't interact with `MenuFlyoutPresenter`. Perhaps throwing a `NotSupportedException` or logging a warning might be better than just an ignore comment though.

File: src/Avalonia.Controls/Flyouts/MenuFlyout.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/MenuFlyout.cs:15]: I think you're right. I forgot to check.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:231]: I'll add some comments to help understanding. The `TransientWithDismissOnPointerMoveAway` is mostly a pointer thing, and AFAICT has no keyboard support in UWP (b/c the Popup is never focused). Although b/c `DefaultMenuInteractionHandler` handles most interactions in `MenuFlyout` I did forget keyboard support for `Flyout` so I'll add that in as part of the changes.

File: src/Avalonia.Controls/Flyouts/MenuFlyout.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/MenuFlyout.cs:15]: Actually, the MenuFlyout page of the XamlControlsGallery shows the MenuFlyout in Top mode, so I guess it does still default to top.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:440]: Is there a mapping to the "menu" key in the `Key` enum? It seems like `Key.Apps` is the match if you handle OnKeyDown, but I'm not sure if this is correct. `ContextMenu` didn't handle this, at least I can't find where and the menu key doesn't appear to do anything. Shift+F10 is an alternate way in Windows too, but I don't know about Mac/Linux

Comment (maxkatz6) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:440]: > Shift+F10

Ah, yes, forgot about this one. It should be implemented too.
And yes, it seems Avalonia doesn't have any of them.

In Avalonia we have "platform hotkeys" implementation for different platforms.
I believe "ContextMenu" should be also included in this list, so it can be configured per platform as well.
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Input/Platform/PlatformHotkeyConfiguration.cs
Later specific keydown event args can be matched with these gestures like it was done in TextBox class
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/TextBox.cs#L602

Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:440]: I'm thinking it may be best to hold off for now then (but I agree this should be added), and I'll defer to someone who knows a little more about the cross platform key gestures - I don't want to wire something up incorrectly. Since ContextMenu didn't already do this, the lack of it shouldn't be noticeable for now.

File: src/Avalonia.Controls/Button.cs
Comment (amwx) [reply src/Avalonia.Controls/Button.cs]: Changed to StyledProperty

File: src/Avalonia.Controls/Button.cs
Comment (amwx) [reply src/Avalonia.Controls/Button.cs]: Added property changed handler to `Button.Flyout` to handle if the Flyout is changed (close the old flyout first). If the Flyout itself is opened on another control, that logic was already there.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:474]: Add a check before showing ContextFlyout. If ContextMenu is also set, logs warning and defaults to ContextMenu.

File: src/Avalonia.Controls/Flyouts/FlyoutPresenter.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/FlyoutPresenter.cs:10]: Border.CornerRadius.AddOwner<FlyoutPresenter>();

**Comment by amwx** [comment]
@maxkatz6 I've made the changes, if there's something else, or something I forgot, let me know.

File: src/Avalonia.Themes.Default/FlyoutPresenter.xaml
Comment (maxkatz6) [inline src/Avalonia.Themes.Default/FlyoutPresenter.xaml:13]: For default theme we don't need corner radius to be set on the presenter, but it should be set on the border with TemplateBinding anyway.

File: src/Avalonia.Controls/Flyouts/FlyoutPresenter.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutPresenter.cs]: Good catch. Fixed.

File: src/Avalonia.Themes.Default/FlyoutPresenter.xaml
Comment (amwx) [reply src/Avalonia.Themes.Default/FlyoutPresenter.xaml:13]: Fixed

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (maxkatz6) [inline src/Avalonia.Controls/Flyouts/FlyoutBase.cs:129]: It's probably bad idea to make popup find out by itself where is the pointer located instead of telling him specific location as it was done in UWP. Especially when it should be pretty straightforward to pass pointer location from PointerPressed event to this method.
But it's not something that can be achieved in this PR without big changes in Popup positioning code (I haven't checked it detailed though).
@kekekeks

Comment (maxkatz6) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs:129]: So for now I think it's fine to have this method like it is and deprecated later, when Popup will be able to be positioned on specific pointer position.

**Comment by maxkatz6** [comment]
Looks good to me, but I want somebody else to review code too.
@AvaloniaUI/core

File: src/Avalonia.Controls/Flyouts/MenuFlyout.cs
Comment (MarchingCube) [inline src/Avalonia.Controls/Flyouts/MenuFlyout.cs:68]: Perhaps this can be a static function somewhere in `FlyoutBase` since this is exactly the same as the one in `Flyout`.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (MarchingCube) [inline src/Avalonia.Controls/Flyouts/FlyoutBase.cs:55]: Missing access modifiers.

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (MarchingCube) [inline src/Avalonia.Controls/Flyouts/FlyoutBase.cs:53]: Perhaps this should be a property instead? I haven't seen many (or any) protected fields in this codebase.

**Review by MarchingCube — CHANGES_REQUESTED** [review]
Generally looks fine to me as well.
Are concepts like `MenuFlyoutItem` supposed to be ported eventually too (not in the scope of this PR)?

**Comment by maxkatz6** [comment]
@MarchingCube 
> Are concepts like MenuFlyoutItem supposed to be ported eventually too (not in the scope of this PR)?

It was discussed above, and we most likely will extend current MenuItem API instead of porting MenuFlyoutItem API. But it's not settled yet.

**Comment by MarchingCube** [comment]
@maxkatz6 Thanks, missed that part of the discussion. I wonder if custom items would make implementing menu items hosting interactive content easier. Like I would like to have a structure like:
```
> root:
- > [MenuItem] Element
- - > [MenuItem] Clear element
- - > [HostedMenuItem] Contains a search box and a list of elements. Clicking on an item closes menu. 
```

File: src/Avalonia.Controls/Flyouts/Flyout.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/Flyout.cs]: Fixed

File: src/Avalonia.Controls/Flyouts/MenuFlyout.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/MenuFlyout.cs]: Fixed

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs]: Added

File: src/Avalonia.Controls/Flyouts/FlyoutBase.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/FlyoutBase.cs]: Fixed

File: src/Avalonia.Controls/Flyouts/Flyout.cs
Comment (amwx) [reply src/Avalonia.Controls/Flyouts/Flyout.cs]: Fixed

**Review by maxkatz6 — APPROVED** [review]

**Review by MarchingCube — APPROVED** [review]
LGTM.

**Comment by maxkatz6** [comment]
@amwx I found one issue with flyouts, which works fine in context menu somehow.
Subitems is not closed as expected on macOS.

https://user-images.githubusercontent.com/3163374/112464833-5e9e8a00-8d3a-11eb-946d-96cf2070e54f.mov

**Comment by maxkatz6** [comment]
Same issue can be reproduced on Windows with OverlayPopups=True
image

**Comment by robloo** [comment]
The checkbox in the Menu Item with Checkbox is also cutoff on the right side. Not sure if that issue was introduced by the new flyout presenter style and its padding.

**Comment by amwx** [comment]
@maxkatz6 I think I've fixed the subitem issue. I copied some logic from ContextMenu that explicitly closes subitems when closing the original menu/flyout and I can't reproduce it now (at least on Windows w/ overlay popups).


> 
> 
> The checkbox in the Menu Item with Checkbox is also cutoff on the right side. Not sure if that issue was introduced by the new flyout presenter style and its padding.

No, that's also occurs in the old context menu, so its an issue in the `MenuItem` template

**Comment by Splitwirez** [comment]
@amwx Just wanted to say, I'm really excited to see work being done on these Flyouts :+1:

Also, uh...I figured it might be worth mentioning that I've just reproduced the submenu auto-closing problem a few minutes ago on Kubuntu Linux 20.04 (x64). If a video recording of the problem occurring would be of any use, I can try to get you one.

**Comment by amwx** [comment]
> Also, uh...I figured it might be worth mentioning that I've just reproduced the submenu auto-closing problem a few minutes ago on Kubuntu Linux 20.04 (x64). If a video recording of the problem occurring would be of any use, I can try to get you one.

@Splitwirez  Is it the same problem as above (where the submenu stays open after parent closes)? If different, is this specific to the flyout implementation, or can it be repro'd with existing menus. Yes, a video may help too

**Pull request merged**
